stellargraph/layer/graphsage.py

```diff
@@ -1048,56 +1086,70 @@ class DirectedGraphSAGE(GraphSAGE):
-        input_dim (int): The dimensions of the node features used as input to the model.
-    """
+    Note::
+        If a generator is not specified, then additional keyword arguments must be supplied:
+
+        * in_samples (list): The number of in-node samples per layer in the model.
+
+        * out_samples (list): The number of out-node samples per layer in the model.
+
+        * input_dim (int): The dimensions of the node features used as input to the model.
+
+        * multiplicity (int): The number of nodes to process at a time. This is 1 for a node inference 
+          and 2 for link inference (currently no others are supported).    """
 
     def _get_sizes_from_generator(self, generator):
         """
         Sets in_samples, out_samples and input_feature_size from the generator.
         Args:
              generator: The supplied generator.
         """
-        self.in_samples = generator.generator.in_samples
+
+        if not isinstance(generator, (DirectedGraphSAGENodeGenerator,)):
+            raise TypeError(
+                "Generator should be an instance of DirectedGraphSAGELinkGenerator or DirectedGraphSAGENodeGenerator"
+            )
+
+        self.in_samples = generator.in_samples
         if len(self.in_samples) != self.max_hops:
             raise ValueError(
                 "Mismatched lengths: in-node sample sizes {} versus layer sizes {}".format(
                     self.in_samples, self.layer_sizes
                 )
             )
-        self.out_samples = generator.generator.out_samples
+        self.out_samples = generator.out_samples
         if len(self.out_samples) != self.max_hops:
             raise ValueError(
                 "Mismatched lengths: out-node sample sizes {} versus layer sizes {}".format(
                     self.out_samples, self.layer_sizes
                 )
             )
-        feature_sizes = generator.generator.graph.node_feature_sizes()
+        feature_sizes = generator.graph.node_feature_sizes()
         if len(feature_sizes) > 1:
             raise RuntimeError(
                 "DirectedGraphSAGE called on graph with more than one node type."
             )
         self.input_feature_size = feature_sizes.popitem()[1]
+        self.multiplicity = generator.multiplicity
 
     def _get_sizes_from_keywords(self, **kwargs):
         """
         Sets in_samples, out_samples and input_feature_size from the keywords.
         Args:
              kwargs: The additional keyword arguments.
         """
-        self.in_samples = kwargs.get("in_samples")
-        self.out_samples = kwargs.get("out_samples")
-        self.input_feature_size = kwargs.get("input_dim")
-        if (
-            self.in_samples is None
-            or self.out_samples is None
-            or self.input_feature_size is None
-        ):
-            raise ValueError(
-                "If generator is not provided, in_samples, out_samples and input_dim must be specified."
+        try:
```

Comment:
Interesting pattern. I know there are people who argue stridently against using try/except unless it is absolutely necessary. I guess I lean that way a bit too (hence the way I originally wrote the code), but I'm not fussed. At least your code does make it clear that those parameters are required.

Reply:
Yes, I'm somewhat ambivalent about it myself.  I felt it made things neater here though.